ambry-account/src/main/java/com/github/ambry/account/HelixAccountService.java

```diff
@@ -0,0 +1,492 @@
+        ZNRecord newRecord;
+        if (currentData == null) {
+          logger.debug(
+              "ZNRecord does not exist on path={} in HelixPropertyStore when updating accounts. Creating a new ZNRecord.",
+              FULL_ACCOUNT_METADATA_PATH);
+          newRecord = new ZNRecord(String.valueOf(ZN_RECORD_ID));
+        } else {
+          newRecord = currentData;
+        }
+        Map<String, String> accountMap = newRecord.getMapField(ACCOUNT_METADATA_MAP_KEY);
+        if (accountMap == null) {
+          logger.debug("AccountMap does not exist in ZNRecord when updating accounts. Creating a new accountMap");
+          accountMap = new HashMap<>();
+        }
+        AccountInfoMap remoteAccountInfoMap;
+        try {
+          remoteAccountInfoMap = new AccountInfoMap(accountMap);
+        } catch (JSONException e) {
+          // Do not depend on Helix to log, so log the error message here.
+          logger.error("Exception occurred when building AccountInfoMap from accountMap={}", accountMap, e);
+          accountServiceMetrics.remoteDataCorruptionErrorCount.inc();
+          throw new IllegalStateException("Exception occurred when building AccountInfoMap from accountMap", e);
+        }
+        // if there is any conflict with the existing record, fail the update. Exception thrown in this updater will
+        // be caught by Helix and helixStore#update will return false.
+        if (hasConflictingAccount(accounts, remoteAccountInfoMap)) {
+          String message = "Updating accounts failed because one account to update conflicts with existing accounts";
+          // Do not depend on Helix to log, so log the error message here.
+          logger.error(message);
+          throw new IllegalArgumentException(message);
+        } else {
+          for (Account account : accounts) {
+            try {
+              accountMap.put(String.valueOf(account.getId()), account.toJson().toString());
+            } catch (Exception e) {
+              String message = "Updating accounts failed because unexpected exception occurred when updating accountId="
+                  + account.getId() + " accountName=" + account.getName();
+              // Do not depend on Helix to log, so log the error message here.
+              logger.error(message, e);
+              throw new IllegalStateException(message, e);
+            }
+          }
+          newRecord.setMapField(ACCOUNT_METADATA_MAP_KEY, accountMap);
+          return newRecord;
+        }
+      }, AccessOption.PERSISTENT);
+    }
+    long timeForUpdate = System.currentTimeMillis() - startTimeMs;
+    if (hasSucceeded) {
+      logger.trace("Completed updating accounts, took time={} ms", timeForUpdate);
+      accountServiceMetrics.updateAccountTimeInMs.update(timeForUpdate);
+      // notify account changes after successfully update.
+      if (notifier.publish(ACCOUNT_METADATA_CHANGE_TOPIC, FULL_ACCOUNT_METADATA_CHANGE_MESSAGE)) {
```

Comment:
Is this notify just so that AccountService can react to account updates more quickly, or is there a race condition that this addresses where the helix listener would not convey the message?